src/main/kotlin/jp/atcoder/library/kotlin/stringAlgorithm/StringAlgorithm.kt

```diff
@@ -0,0 +1,265 @@
+package jp.atcoder.library.kotlin.stringAlgorithm
+
+import java.util.*
+import java.util.function.Consumer
+
+
+object StringAlgorithm {
+    private fun saNaive(s: IntArray): IntArray {
+        val n = s.size
+        val sb = arrayOfNulls<Int>(n)
+        for (i in 0 until n) {
+            sb[i] = i
+        }
+        Arrays.sort(sb) { l: Int?, r: Int? ->
+            var vl = l
+            var vr = r
+            while (vl!! < n && vr!! < n) {
+                if (s[vl] != s[vr]) return@sort s[vl] - s[vr]
+                vl++
+                vr++
+            }
+            -(vl - vr!!)
+        }
+        val sa = IntArray(n)
+        for (i in 0 until n) {
+            sa[i] = sb[i]!!
+        }
+        return sa
+    }
+
+    private fun saDoubling(s: IntArray): IntArray {
+        val n = s.size
+        val sb = arrayOfNulls<Int>(n)
```

Comment:
見てみたところJava版からそうなのですが、primitive型配列を使用するようにしたほうが速度的によいですね。
Kotlin版では定数倍は一旦許容するとしてるのですが、Java ACLから遅いのであればそっちから改善したほうがいいかも…一旦マージするのであとでJava版のほう見てみます。

[TIPS] 
Kotlinの`Int?`と`Int`は、Javaでいう`Integer`と`int`なので、ボクシングのコストが掛かってしまいます。
なので、使えるところであればprimitive型配列である`IntArray`のほうが望ましいです。

また、これも機能的には全く同じなのですが`sb[i] = i`の初期化であれば、for文を使わずとも下記のように書けるので個人的には好みですね。
```suggestion
        val sb = IntArray(n){ it }
```

Reply:
Java版の方ですでに同様の検討をしているみたいでした。優秀すぎる…
https://github.com/NASU41/AtCoderLibraryForJava/issues/4

Reply:
私もJavaのライブラリは全部プリミティブ配列のもの使ってますね。Javaは意外とプリミティブ型フレンドリーなのでそういう人多そうですが、元のACLがSTLをばりばり使ってしまってるため直移植してIntegerになってる予感。